File: Javascript/Sorting/quick_sort.js
/**
 * This file is part of [Algorithms]
 *
 * (c) 2024 [Nainik Mehta] <[nainikmehta25@gmail.com]>
 *
 * --------------------------------------------------
 *
 * @module app.v1.quickSort
 * @description Quick Sort is a divide-and-conquer sorting algorithm that works by selecting a 'pivot' element from the array and partitioning the other elements into two sub-arrays according to whether they are less than or greater than the pivot. The sub-arrays are then recursively sorted. This process continues until the entire array is sorted. It is a highly efficient sorting algorithm with an average time complexity of O(n log n).
 * @author [Nainik Mehta] <[https://www.linkedin.com/in/nainik-mehta-25nk12/]>
 * @github - Nainik2509  <[https://github.com/Nainik2509/]>
 * @version 1.0.0
 *
 * --------------------------------------------------
 */

function quickSort(arr) {
  if (arr.length <= 1) {
    return arr; // Base case: array with 0 or 1 elements is already sorted
  }

  const pivot = arr[0]; // Select the first element as the pivot
  const left = []; // Array to store elements less than the pivot
  const right = []; // Array to store elements greater than or equal to the pivot

  // Partition the array into left and right sub-arrays
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] < pivot) {
      left.push(arr[i]); // Add elements less than pivot to the left sub-array
    } else {
      right.push(arr[i]); // Add elements greater than or equal to pivot to the right sub-array
    }
  }

  // Recursively apply quickSort to the left and right sub-arrays, and concatenate them with the pivot
  return [...quickSort(left), pivot, ...quickSort(right)];
}

// Example usage
const arr = [51, 45, 63, 71, 43, 23, 90];
console.log('Original Array:', arr);
const sortedArr = quickSort(arr);
console.log('Sorted Array:', sortedArr);

File: C++/Sorting/quick_sort.cpp
#include <iostream>
using namespace std;

// Function to swap two elements
void swap(int *x, int *y)
{
    int temp;
    temp = *x;
    *x = *y;
    *y = temp;
}

// Function to partition the array
int partition(int arr[], int low, int high)
{
    int pivot = arr[low]; // Choosing the pivot element
    int i = low;          // Initialize the left pointer
    int j = high;         // Initialize the right pointer

    // Perform partitioning
    do
    {
        // Move the left pointer until an element greater than the pivot is found
        do
        {
            i++;
        } while (arr[i] <= pivot);

        // Move the right pointer until an element smaller than or equal to the pivot is found
        do
        {
            j--;
        } while (arr[j] > pivot);

        // Swap the elements if the left pointer is less than the right pointer
        if (i < j)
            swap(&arr[i], &arr[j]);
    } while (i < j);

    // Swap the pivot element with the element at the right pointer
    swap(&arr[j], &arr[low]);

    // Return the index of the pivot element
    return j;
}

// Function to perform Quick Sort
void QuickSort(int arr[], int low, int high)
{
    if (low < high)
    {
        // Partition the array
        int p = partition(arr, low, high);

        // Recursively sort the left and right subarrays
        QuickSort(arr, low, p);
        QuickSort(arr, p + 1, high);
    }
}

int main()
{
    // Input array
    int arr[] = {51, 45, 63, 71, 43, 23, 90};
    int n = sizeof(arr) / sizeof(arr[0]); // Calculate the size of the array
    QuickSort(arr, 0, n - 1);             // Perform Quick Sort
    int i;
    for (i = 0; i < n; i++)
    {
        cout << arr[i] << " "; // Print the sorted array
    }
}
